### PR TITLE
Fixes #39. Implement basic OAuth2 support

### LINKED ISSUE #39: simple oauth2 server side support
For 3.2 we would like to have some simple oauth2 support, the MVP is to support authentication using a third party such as Github/Facebook/Google

### PR DESCRIPTION
**Pull request by pmlopes**

### CODE REVIEW

File: vertx-auth-oauth2/src/test/java/io/vertx/ext/auth/test/oauth2/OAuth2AccessTokenTest.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+import io.vertx.test.core.VertxTestBase;
+import org.junit.Test;
+
+import java.util.concurrent.CountDownLatch;
+
+import static io.vertx.ext.auth.oauth2.impl.OAuth2API.*;
+
+public class OAuth2AccessTokenTest extends VertxTestBase {
+
+  private static final JsonObject fixture = new JsonObject(
```
Comment (cescoffier) [inline vertx-auth-oauth2/src/test/java/io/vertx/ext/auth/test/oauth2/OAuth2AccessTokenTest.java]: is it used ?

File: vertx-auth-oauth2/src/test/java/io/vertx/ext/auth/test/oauth2/OAuth2FailureTest.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+import org.junit.Test;
+
+import java.net.URLEncoder;
+import java.util.concurrent.CountDownLatch;
+
+import static io.vertx.ext.auth.oauth2.impl.OAuth2API.stringify;
+
+public class OAuth2FailureTest extends VertxTestBase {
+
+  private static final JsonObject fixture = new JsonObject(
```
Comment (cescoffier) [inline vertx-auth-oauth2/src/test/java/io/vertx/ext/auth/test/oauth2/OAuth2FailureTest.java]: Is it used ?

File: vertx-auth-oauth2/src/test/java/io/vertx/ext/auth/test/oauth2/OAuth2FailureTest.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      .put("redirect_uri", "http://callback.com");
+
+  private static final JsonObject oauthConfig = new JsonObject()
+      .put("code", "code")
+      .put("redirect_uri", "http://callback.com")
+      .put("client_secret", "client-secret")
+      .put("grant_type", "authorization_code")
+      .put("client_id", "client-id");
+
+  private static final JsonObject authorizeConfig = new JsonObject()
```
Comment (cescoffier) [inline vertx-auth-oauth2/src/test/java/io/vertx/ext/auth/test/oauth2/OAuth2FailureTest.java]: is it used ?

File: vertx-auth-oauth2/src/main/asciidoc/groovy/index.adoc [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+  }
+})
+
+----
+
+==== Authorization Code flow
+
+The Authorization Code flow is made up from two parts. At first your application asks to the user the permission to
+access their data. If the user approves the OAuth2 server sends to the client an authorization code. In the second
+part, the client POST the authorization code along with its client secret to the Lelylan in order to get the access
```
Comment (cescoffier) [inline vertx-auth-oauth2/src/main/asciidoc/groovy/index.adoc]: I'm surprised to see a reference to Lelylan here. Did you forget to edit the sentence ?

File: vertx-auth-oauth2/src/test/java/io/vertx/ext/auth/test/oauth2/OAuth2FailureTest.java (hunk 2 above)
Comment (pmlopes) [inline vertx-auth-oauth2/src/test/java/io/vertx/ext/auth/test/oauth2/OAuth2FailureTest.java]: nope, too much c&p

File: vertx-auth-oauth2/src/test/java/io/vertx/ext/auth/test/oauth2/OAuth2FailureTest.java (hunk 3 above)
Comment (pmlopes) [inline vertx-auth-oauth2/src/test/java/io/vertx/ext/auth/test/oauth2/OAuth2FailureTest.java]: nope, too much c&p

**Comment by cescoffier** [comment]
It looks good. It would be nice to extend the documentation with the configuration for the main Oauth provider and a link on the page explaining how they register their application.

Main providers are for me:
github, twitter, facebook, google, and maybe linkedin.

**Comment by pmlopes** [comment]
that documentation i've it on vertx-web there i've documented:
- github
- google
- linkedin

**Comment by pmlopes** [comment]
i can try to verify the configuration for twitter and maybe facebook too... i just need to create an account there ;P

**Comment by pmlopes** [comment]
documented the major providers

**Pull request merged**